examples/keyvalue-inmemory/main.go

```diff
@@ -69,12 +71,24 @@ func run() error {
 
 func (p *Provider) handleNewSourceLink(link provider.InterfaceLinkDefinition) error {
 	log.Println("Handling new source link", link)
+	err := p.establishSourceLink(link)
+	if err != nil {
+		log.Println("Failed to establish source link", link, err)
+		p.failedSourceLinks[link.Target] = link
```

Comment:
What do you think about moving all of the establishing link functionality into `link.go`, keeping `main.go` for just binary setup, `provider.go` for implementing the interface, and then `link.go` for linking logic?

I'm sure there's many ways we could restructure the example, so putting a good organization forward to start with is a good way to help out new folks to the SDK 😄 Curious to hear your thoughts

Reply:
> What do you think about moving all of the establishing link functionality into `link.go`, keeping `main.go` for just binary setup, `provider.go` for implementing the interface, and then `link.go` for linking logic?
> 
> I'm sure there's many ways we could restructure the example, so putting a good organization forward to start with is a good way to help out new folks to the SDK 😄 Curious to hear your thoughts

Yes, this makes sense to me and is a suggested Go approach. Would you like this refactor to be a part of this ticket, might be a good one for another issue for me to work on!

Reply:
A part of this ticket would be great I think! As long as you don't mind pushing a few more commits

Reply:
I'll layer in a few more commits with these updates and request a re-review when it's ready!